### PR TITLE
Implement `groupby::merge` for `collect_list` and `collect_set`

### LINKED ISSUE #7839: [FEA] Support `MERGE_LISTS` and `MERGE_SETS` for `groupby::aggregate`
**Is your feature request related to a problem? Please describe.**
To perform collect aggregations in Spark, (like other aggregation functions) it needs to go through two phases. In phase one, we perform collect aggregation within each node locally, which can be implemented by cuDF group_by with AggOp `collect_list` or `collect_set`. In phase two, we need to merge partial aggregation results from each node. In terms of collect aggregations, we concatenate multiple lists/sets into one. For instance,
```
Node A has a partial result of collect_list in group_by: [('A', [1, 2, 3]), ('B', [10, 20])]
Node B has a result: [('B', [30, 40]), ('C', [100, 200])]
Node C has a result: [('A', [4]), ('C', [300, 400])]

For phase two, we have a combined input table: [('A', [1, 2, 3]), ('B', [10, 20]), ('B', [30, 40]), ('C', [100, 200]), ('A', [4]), ('C', [300, 400])]
Then, we concatenate lists sharing same group key and we suppose to get the result:  [('A', [1, 2, 3, 4]), ('B', [10, 20, 30, 40]), ('C', [100, 200, 300, 400])]
```
I think we need two additional AggOps (`concatenate_list` and `concatenate_set`) for current feature.

### PR DESCRIPTION
**Pull request by ttnghia**
Groupby aggregations can be performed for distributed computing by the following approach:
 * Divide the dataset into batches
 * Run separate (distributed) aggregations on those batches
 * Merge the results of the step above into one final result

This PR supports merging operations for `collect_list` and `collect_set`.

Closes #7839.

### PS: This is a WIP and not ready for review.

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -195,6 +195,7 @@ add_library(cudf
     src/groupby/sort/group_argmin.cu
     src/groupby/sort/aggregate.cpp
     src/groupby/sort/group_collect.cu
+    src/groupby/sort/group_collect_merge.cu
```
Comment (robertmaynard) [inline cpp/CMakeLists.txt:198]: I don't see any direct kernel launches in `group_collect_merge.cu` can this file instead by c++ file?

Comment (ttnghia) [reply cpp/CMakeLists.txt:198]: Thanks for reminding me about this, Robert. This is WIP and not complete yet, I want to have a quick working prototype for other people in my team to have a look and I will add more implementation to this file later.

File: cpp/src/groupby/hash/groupby.cu [hunk 2]
```diff
@@ -631,6 +618,16 @@ std::unique_ptr<table> groupby_null_templated(table_view const& keys,
...
 
+/**
+ * @brief Indicates whether the specified aggregation operation can be computed
+ * with a hash-based implementation.
+ *
+ * @param t The aggregation operation to verify
+ * @return true `t` is valid for a hash based groupby
+ * @return false `t` is invalid for a hash based groupby
+ */
+bool is_hash_aggregation(aggregation::Kind t) { return array_contains(hash_aggregations, t); }
```
Comment (jrhemstad) [inline cpp/src/groupby/hash/groupby.cu:629]: Separating the definition makes this no longer `constexpr`. If you want it available in the header, move it's definition there as well and keep it `constexpr`.

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by sperlingxx** [comment]
The implementation looks good. But I have some concern about the interface: how can we perform this mergeOp along with other AggOps (like Sum/Count/Max) ?  

In the JNI wrapper of `groupByAggregate`, we combine multiple aggregation requests with the help of `groupBy::aggregate` ([code link]()).  Perhaps we can separate all merge requests from other requests, and handle them separately with `groupBy::merge`. Then we combine the results of `groupBy::aggregate` and `groupBy::merge`. I am not sure it is a good approach.

In addition, if this merge API is specialized for spark-rapids, the concatenation of multiple keys and values looks unnecessary. Because spark-rapids will concatenate all partial results (also with `cudf::concatenate`) before the phase 2 aggregation (merging partial results). [code link](https://github.com/NVIDIA/spark-rapids/blob/branch-21.06/sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala#L377)
Therefore, in terms of spark-rapids, it looks better if we can run mergeOp with `groupBy::aggregate` instead of `groupBy::merge`.

**Comment by revans2** [comment]
> The implementation looks good. But I have some concern about the interface: how can we perform this mergeOp along with other AggOps (like Sum/Count/Max) ?

I totally agree this is why I wanted just a concat aggregation.  Not some new merge op API. Just a concat that is a regualar aggregation, like collect list or collect set.

**Comment by ttnghia** [comment]
> > The implementation looks good. But I have some concern about the interface: how can we perform this mergeOp along with other AggOps (like Sum/Count/Max) ?
> 
> I totally agree this is why I wanted just a concat aggregation. Not some new merge op API. Just a concat that is a regualar aggregation, like collect list or collect set.

Initially I did it. But then I found out that the new API will need to operate on (merge) multiple pair of keys-values (which are the grouped keys and lists columns resulted from the previous `collect_list`), and those results may have very different numbers of rows. Since `groupby::aggregate` requires all the input requests to have the same number of rows, I decided to resort this out to a new `groupby::merge` API.

Do you guys have any suggestion for this issue (mismatching numbers of input rows of the `collect_list` results for merging)?

**Comment by revans2** [comment]
To merge the outputs we concat them into a single table and then call another aggregation on them.  This is actually very simplified because it involves multiple machines shuffling the data around.  We don't need a new merge API.  Just concat_list and concat_set aggregations. You don't need to worry about how the data gets chopped up and redistributed. You just need to worry about adding in the desired aggregations we will handle the rest.  We already do.

**Comment by jlowe** [comment]
> Do you guys have any suggestion for this issue (mismatching numbers of input rows of the collect_list results for merging)?

It's easier to think of this as it's own, standalone type of aggregation, i.e.: given a table, aggregate the list columns associated with a particular key by concatenating the list values together into a new list (and removing duplicates if it's collect_set merge).

As @sperlingxx said,
> the concatenation of multiple keys and values looks unnecessary. Because spark-rapids will concatenate all partial results (also with cudf::concatenate) before the phase 2 aggregation (merging partial results).

Therefore there's no need to consider multiple tables or multiple `collect_list` results.  It's just an aggregation like any other aggregation.  Therefore there's no issue with mismatched result row counts.  For this aggregation, there's just one input table.  There are some key columns and a list column, the latter which will be aggregated with this new concat_lists aggregation (or whatever we want to call the aggregation operation).  There's no "mismatch of row counts" because there's just keys and lists associated with keys.  Sort the table by the key columns and concatenate all the sequential rows associated with the same key.  There may be different numbers of rows associated for one key than another key, but that's no different than what collect_list or other aggregations need to handle.

**Comment by ttnghia** [comment]
Got it. Thanks all, I'm starting a new PR according to your suggestions.

**Pull request closed without merging**